File: sort/75.c
//当遇到0或2时, 把这个数放到前或者后的正确位置
void swap(int* a, int* b)
{
    int temp;
    temp = *a;
    *a = *b;
    *b = temp;
}
void sortColors(int* nums, int numsSize) {
    int lo = 0;
    int hi = numsSize - 1;

    int i = 0;
    while(i < numsSize)
    {
        if(nums[i] == 0 && lo < i)
        {
            swap(&nums[i], &nums[lo]);
            lo++;
            continue;
        }
        if(nums[i] == 2 && i < hi)
        {
            swap(&nums[i], &nums[hi]);
            hi--;
            continue;
        }
        i++;
    }
}

File: two_pointers/75.go
func sortColors(nums []int)  {
    var red, white, blue int
    blue = len(nums) - 1
    for white <= blue {
        if nums[white] == 0 {
            nums[white], nums[red] = nums[red], nums[white]
            red++
            // 此时可以移动white 因为换过来的必然是1 在white坐标之前的值 都不可能是2
            white++
        }else if nums[white] == 1 {
            white++
        }else {
            nums[white], nums[blue] = nums[blue], nums[white]
            blue--
            // 此时不可以移动white 因为换过来的还可能是2
        }
    }
}
